### PR TITLE
[QEff. Finetune] : Add Initial Phase-1 Support in QEff.

### PR DESCRIPTION
**Pull request by quic-mamta**
Here's a summary for Finetune PR:

---

### Summary

**Added Support for PEFT LoRA Finetuning in Cloud AI 100 Stack**

- **Initial Support**: 
  - Implemented with QAIC PyTorch eager support.
  - Restricted to four datasets: Samsum, Grammar, GSM8K, and Alpaca.
  - Detailed documentation provided for adding custom dataset pipeline support.
  - The Foundational model for this phase will be [**Llama3.2 1B**](https://huggingface.co/meta-llama/Llama-3.2-1B)

- **Usage**:
  - CLI option available in `QEfficient.cloud.finetune`.
  - More details in `finetune.md` under `docs`.

- **Multi-Device Finetuning**:
  - Added support using Distributed Data Parallel (DDP).

- **Verification and Evaluation**:
  - Scripts can be used to verify runs and evaluate perplexity on QAIC and GPU.
  - The finetuned checkpoints can be run with QEff. AOT infer stack, completing its training to deployment cycle. (*Provided we support the model architecture in QEff.)

---

### CODE REVIEW

File: QEfficient/cloud/finetune.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        getattr(torch, torch_device.type).set_device(dist.get_rank())
+
+    # Set the seeds for reproducibility
+    torch.manual_seed(train_config.seed)
+    random.seed(train_config.seed)
+    np.random.seed(train_config.seed)
+
+    # Load the pre-trained model and setup its configuration
+    # config = AutoConfig.from_pretrained(train_config.model_name)
+    model = AutoModelForCausalLM.from_pretrained(
```
Comment (anujgupt-github) [inline QEfficient/cloud/finetune.py:71]: Question: 
Is it better to load model via QEffAutoModelForCausalLM (and in turn have all AI100 specific changes within) and then finetune, or load via default AutoModelForCausalLM, finetune, and then do AI100 specific changes for inference?

File: QEfficient/cloud/finetune.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    if train_config.use_peft:
+        # Load the pre-trained peft model checkpoint and setup its configuration
+        if train_config.from_peft_checkpoint:
+            model = PeftModel.from_pretrained(model, train_config.from_peft_checkpoint, is_trainable=True)
+            peft_config = model.peft_config
+        # Generate the peft config and start fine-tuning from original model
+        else:
+            peft_config = generate_peft_config(train_config, kwargs)
+            model = get_peft_model(model, peft_config)
+        model.print_trainable_parameters()
```
Comment (anujgupt-github) [inline QEfficient/cloud/finetune.py:105]: what is the expected behavior when this is false?

File: QEfficient/cloud/finetune.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    model.to(train_config.device)
+    optimizer = optim.AdamW(
+        model.parameters(),
+        lr=train_config.lr,
+        weight_decay=train_config.weight_decay,
+    )
+    scheduler = StepLR(optimizer, step_size=1, gamma=train_config.gamma)
+
+    # wrap model with DDP
+    if train_config.enable_ddp:
```
Comment (anujgupt-github) [inline QEfficient/cloud/finetune.py:181]: we should shuffle dataset if ddp is enabled

File: QEfficient/finetune/configs/peft_config.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    lora_alpha: int = 32
+    target_modules: List[str] = field(default_factory=lambda: ["q_proj", "v_proj"])
+    bias = "none"
+    task_type: str = "CAUSAL_LM"
+    lora_dropout: float = 0.05
+    inference_mode: bool = False
+
+
+@dataclass
+class llama_adapter_config:
```
Comment (anujgupt-github) [inline QEfficient/finetune/configs/peft_config.py:27]: why do we call it llama_adapter_config?

File: QEfficient/finetune/summarize_infer.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+with torch.inference_mode():
+    print(
+        tokenizer.decode(
+            model.generate(**model_input, max_new_tokens=50, do_sample=False)[0],
+            skip_special_tokens=True,
+        )
+    )
+
+save_dir = "meta-llama-samsum/trained_weights/step_2000"
```
Comment (anujgupt-github) [inline QEfficient/finetune/summarize_infer.py:67]: Why is this hardcoded?

File: QEfficient/finetune/summarize_infer.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
+# -----------------------------------------------------------------------------
```
Comment (anujgupt-github) [inline QEfficient/finetune/summarize_infer.py:1]: The filename _infer.py doesn't sit well with what we are trying to do here?

File: docs/source/finetune.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
+# Finetune Infra
+
+This repository provides the infrastructure for finetuning models using different hardware accelerators such as QAIC, CUDA, and CPU.
```
Comment (anujgupt-github) [inline docs/source/finetune.md:3]: Remove references to CUDA.
We can mention that the same CLI can be used to run FT on GPU by setting the device as ...

File: QEfficient/cloud/finetune.py (hunk 1 above)
Comment (vbaddi) [reply QEfficient/cloud/finetune.py:71]: We can do this in inference stage, as soon as we wrap in QEff model, structure changes due to AOT transforms.
We will do this in when we want to have e2e deployment and infer using the AOT stack.

File: QEfficient/finetune/summarize_infer.py (hunk 6 above)
Comment (vbaddi) [reply QEfficient/finetune/summarize_infer.py:1]: nit: this goes away, the finetuned checkpoints are directly integrated with QEff. inference stack

File: QEfficient/finetune/summarize_infer.py (hunk 5 above)
Comment (quic-mamta) [reply QEfficient/finetune/summarize_infer.py:67]: addressed this.

File: QEfficient/finetune/configs/peft_config.py (hunk 4 above)
Comment (quic-mamta) [reply QEfficient/finetune/configs/peft_config.py]: Removed this.

File: QEfficient/cloud/finetune.py (hunk 3 above)
Comment (quic-mamta) [reply QEfficient/cloud/finetune.py:191]: addressed this.

File: scripts/run_ft_model.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
+# -----------------------------------------------------------------------------
```
Comment (vbaddi) [inline scripts/run_ft_model.py:1]: nit: push it inside scripts/finetune.

Comment (quic-mamta) [reply scripts/run_ft_model.py]: done.

**Review by irajagop — APPROVED** [review]

**Review by vbaddi — APPROVED** [review]
